File: commons/java/Answer.java
package answer;

import lombok.experimental.UtilityClass;

@UtilityClass
public class Answer {

    public static void part1(int expected, int result) {
        part(1, expected, result);
    }

    public static void part2(int expected, int result) {
        part(2, expected, result);
    }

    private static void part(int part, int expected, int result) {
        if (expected != result) {
            throw new RuntimeException(String.format(
                "Part %d incorrect, expected %d but got %d", part, expected, result
            ));
        }
        System.out.printf("Part %d: %d \n", part, result);
    }
}

File: commons/rust/answer.rs
use std::cmp::PartialEq;
use std::fmt::Debug;

pub trait Solution: Debug + PartialEq + ToString {}
impl<T: Debug + PartialEq + ToString> Solution for T {}

pub fn part1<T: Solution>(expected: T, result: T) {
    part(1, expected, result);
}

pub fn part2<T: Solution>(expected: T, result: T) {
    part(2, expected, result);
}

fn part<T: Solution>(part: i64, expected: T, result: T) {
    if expected != result {
        panic!(
            "Part {part} incorrect, expected {0} but got {1}",
            expected.to_string(),
            result.to_string(),
        );
    }
    println!("Part {part}: {0}", result.to_string());
}
